File: spikes/010-go-or-rust/bugs/A-omitted-case/go/main.go
// BUG A-1, Go: the switch omits `expiry_reversal`. There is no `default`,
// and the function returns named zero values.
package main

import "fmt"

type AuthEventKind string

const (
	KindAuthorization  AuthEventKind = "authorization"
	KindIncremental    AuthEventKind = "incremental"
	KindAdvice         AuthEventKind = "advice"
	KindReversal       AuthEventKind = "reversal"
	KindClearing       AuthEventKind = "clearing"
	KindExpiry         AuthEventKind = "expiry"
	KindExpiryReversal AuthEventKind = "expiry_reversal"
)

type Effect struct {
	Delta            int64
	IncreaseSide     bool
	ClearsExpiryFlag bool
}

func Normalise(kind AuthEventKind, wire int64) (eff Effect, err error) {
	switch kind {
	case KindAuthorization, KindIncremental:
		return Effect{Delta: wire, IncreaseSide: true}, nil
	case KindClearing, KindReversal:
		return Effect{Delta: wire}, nil
	case KindAdvice:
		return Effect{Delta: wire}, nil
	case KindExpiry:
		return Effect{}, fmt.Errorf("expiry is a flag")
		// case KindExpiryReversal: <-- OMITTED
	}
	return // <-- silent: Effect{}, nil
}

func main() {
	e, err := Normalise(KindExpiryReversal, 0)
	fmt.Printf("go   Normalise(expiry_reversal) = %+v, err=%v\n", e, err)
	fmt.Printf("go   the flag that should have been cleared: ClearsExpiryFlag=%v\n", e.ClearsExpiryFlag)
}

File: spikes/010-go-or-rust/bugs/A-omitted-case/rs/src/main.rs
// BUG A-1, Rust: the match omits `ExpiryReversal`, exactly as the Go does.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AuthEventKind {
    Authorization, Incremental, Advice, Reversal, Clearing, Expiry, ExpiryReversal,
}

#[derive(Debug, Default, Clone, Copy)]
pub struct Effect { pub delta: i64, pub increase_side: bool, pub clears_expiry_flag: bool }

pub fn normalise(kind: AuthEventKind, wire: i64) -> Result<Effect, String> {
    use AuthEventKind::*;
    match kind {
        Authorization | Incremental => Ok(Effect { delta: wire, increase_side: true, ..Default::default() }),
        Clearing | Reversal => Ok(Effect { delta: wire, ..Default::default() }),
        Advice => Ok(Effect { delta: wire, ..Default::default() }),
        Expiry => Err("expiry is a flag".into()),
        // ExpiryReversal => <-- OMITTED
    }
}

fn main() {
    println!("{:?}", normalise(AuthEventKind::ExpiryReversal, 0));
}
